### PR TITLE
Primary keys may not be unique for variable length regexes

### LINKED ISSUE #2116: Primary keys may not be unique for variable length regexes
### Environment Details
* SDV version: 1.14.0

### Error Description
If a regex has variable length (eg. 3 or 4 characters long with a format such as `'[0-9]{3,4}'`), the primary keys that SDV generates may not always be unique. This is supported by the diagnostic report being less than 100%.

### Steps to reproduce
```python
import pandas as pd

from sdv.metadata import SingleTableMetadata
from sdv.single_table import GaussianCopulaSynthesizer
from rdt.transformers.text import RegexGenerator
from sdv.evaluation.single_table import run_diagnostic

data = pd.DataFrame(data={
    'id': [123, 456, 1234, , 5678],
    'col': [23.4, 12.3, 10.0, 90.4, 10.3]
})

metadata = SingleTableMetadata.load_from_dict({
    'columns': {
        'id': { 'sdtype': 'id', 'regex_format': '[0-9]{3,4}' },
        'col': { 'sdtype': 'numerical' }
    },
    'primary_key': 'id'
})

synth = GaussianCopulaSynthesizer(metadata)

# I am specifically assigning the RegexGenerator for illustration purposes
synth.auto_assign_transformers(data)
synth.update_transformers({
    'id': RegexGenerator(regex_format='[0-9]{3,4}', enforce_uniqueness=True, generation_order='alphanumeric')
})

synth.fit(data)
synthetic_data = synth.sample(num_rows=10_000)

diagnostic = run_diagnostic(data, synthetic_data, metadata)
diagnostic.get_details('Data Validity')
```
Output:
image

### Expected Fix
This case can only happen when:

- The real data has a primary key (or alternate key) stored as an `int` AND
- The metadata for it includes a Regex AND
- The Regex allows for a `'0'` to be in the first position of the string

If _and only if_ all 3 conditions are met, the synthesizer should throw an error.

```python
synthesizer = HSASynthesizer(metadata)
synthesiser.fit(data)
```
```
SynthesizerInputError: Primary key for table 'users' is stored as an int but the Regex allows it to start with '0'. Please remove the Regex or update it to correspond to valid ints.
```

### PR DESCRIPTION
**Pull request by amontanez24**
resolves #2116 
CU-86b17dk49

### CODE REVIEW

**Comment by sdv-team** [comment]
Task linked: [CU-86b17dk49 SDV - Primary keys may not be unique for variable length regexes #2116](https://app.clickup.com/t/86b17dk49)

File: tests/unit/single_table/test_base.py [hunk 1]
```diff
@@ -600,6 +600,47 @@ def test_validate_raises_invalid_data_for_metadata(self):
         instance._validate_constraints.assert_called_once_with(data)
         instance._validate.assert_not_called()
 
+    def test_validate_int_primary_key_regex_starts_with_0(self):
```
Comment (pvk-developer) [inline tests/unit/single_table/test_base.py:603]: can we change `0` to `zero`

File: sdv/multi_table/base.py [hunk 2]
```diff
@@ -363,9 +364,17 @@ def preprocess(self, data):
...
+                synthesizer = self._table_synthesizers[table_name]
+                self._assign_table_transformers(synthesizer, table_name, table_data)
+                processed_data[table_name] = synthesizer._preprocess(table_data)
+            except SynthesizerInputError as e:
+                if INT_REGEX_ZERO_ERROR_MESSAGE in str(e):
+                    raise SynthesizerInputError(
+                        f'Primary key for table "{table_name}" {INT_REGEX_ZERO_ERROR_MESSAGE}'
+                    )
+
+                raise e
```
Comment (pvk-developer) [inline sdv/multi_table/base.py:377]: Why do we need to do the validation here?
Shouldn't this already occur during `self.validate(data)` ?

Comment (amontanez24) [reply sdv/multi_table/base.py:377]: It was requested that in the multi table case we add the table name to the error message

**Review by rwedge — APPROVED** [review]

**Review by pvk-developer — APPROVED** [review]

**Pull request merged**